File: frontend/lib/models/shelter.dart
class Shelter {
  final String id;
  final String name;
  final double lat;
  final double lng;
  final int capacity;
  final int currentOccupancy;
  final double distanceM;
  final String? type;

  Shelter({
    required this.id,
    required this.name,
    required this.lat,
    required this.lng,
    required this.capacity,
    required this.currentOccupancy,
    required this.distanceM,
    this.type,
  });

  factory Shelter.fromJson(Map<String, dynamic> json) {
    return Shelter(
      id: json['id']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      lat: double.tryParse((json['lat'] ?? 0).toString()) ?? 0,
      lng: double.tryParse((json['lng'] ?? 0).toString()) ?? 0,
      capacity: int.tryParse((json['capacity'] ?? 0).toString()) ?? 0,
      currentOccupancy:
          int.tryParse((json['current_occupancy'] ?? 0).toString()) ?? 0,
      distanceM: double.tryParse((json['distance_m'] ?? 0).toString()) ?? 0,
      type: json['type'].toString(),
    );
  }
}

File: frontend-rn/src/types/shelter.ts
export interface Shelter {
  id: string;
  name: string;
  lat: number;
  lng: number;
  capacity: number;
  currentOccupancy: number;
  distanceM: number;
  type?: string;
}

const num = (v: unknown, d = 0) => {
  const n = Number(v);
  return Number.isFinite(n) ? n : d;
};

export function shelterFromJson(j: any): Shelter {
  return {
    id: String(j.id ?? ''),
    name: String(j.name ?? ''),
    lat: num(j.lat),
    lng: num(j.lng),
    capacity: num(j.capacity),
    currentOccupancy: num(j.current_occupancy),
    distanceM: num(j.distance_m),
    type: j.type != null ? String(j.type) : undefined,
  };
}
